File: app/src/main/java/com/bubelov/coins/data/repository/placecategory/PlaceCategoriesDataSourceDb.java
package com.bubelov.coins.data.repository.placecategory;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import com.bubelov.coins.data.database.DbContract;
import com.bubelov.coins.domain.PlaceCategory;

import javax.inject.Inject;
import javax.inject.Singleton;

/**
 * @author Igor Bubelov
 */

@Singleton
public class PlaceCategoriesDataSourceDb implements PlaceCategoriesDataSource {
    private final SQLiteDatabase db;

    @Inject
    PlaceCategoriesDataSourceDb(SQLiteDatabase db) {
        this.db = db;
    }

    @Override
    public PlaceCategory getPlaceCategory(long id) {
        try (Cursor cursor = db.query(DbContract.PlaceCategories.TABLE_NAME,
                null,
                "_id = ?",
                new String[]{String.valueOf(id)},
                null,
                null,
                null)) {
            if (cursor.getCount() == 0) {
                return null;
            } else {
                cursor.moveToFirst();
                return PlaceCategory.builder()
                        .id(cursor.getLong(cursor.getColumnIndex(DbContract.PlaceCategories._ID)))
                        .name(cursor.getString(cursor.getColumnIndex(DbContract.PlaceCategories.NAME)))
                        .build();
            }
        }
    }

    public void addPlaceCategory(PlaceCategory category) {
        ContentValues values = new ContentValues();
        values.put(DbContract.PlaceCategories._ID, category.id());
        values.put(DbContract.PlaceCategories.NAME, category.name());

        db.insertWithOnConflict(DbContract.PlaceCategories.TABLE_NAME,
                null,
                values,
                SQLiteDatabase.CONFLICT_REPLACE);
    }
}

File: app/src/main/java/com/bubelov/coins/repository/placecategory/PlaceCategoriesDataSourceDb.kt
package com.bubelov.coins.repository.placecategory

import android.content.ContentValues
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase

import com.bubelov.coins.database.DbContract
import com.bubelov.coins.model.PlaceCategory

import javax.inject.Inject
import javax.inject.Singleton

/**
 * @author Igor Bubelov
 */

@Singleton
class PlaceCategoriesDataSourceDb @Inject
internal constructor(private val db: SQLiteDatabase) {
    fun getPlaceCategory(id: Long): PlaceCategory? {
        db.query(DbContract.PlaceCategories.TABLE_NAME,
                null,
                "_id = ?",
                arrayOf(id.toString()), null, null, null).use { cursor ->
            return if (cursor.moveToNext()) cursor.toPlaceCategory() else null
        }
    }

    fun addPlaceCategory(category: PlaceCategory) {
        db.insertWithOnConflict(DbContract.PlaceCategories.TABLE_NAME, null,
                category.toContentValues(),
                SQLiteDatabase.CONFLICT_REPLACE)
    }

    private fun PlaceCategory.toContentValues() = ContentValues().apply {
        put(DbContract.PlaceCategories._ID, id)
        put(DbContract.PlaceCategories.NAME, name)
    }

    private fun Cursor.toPlaceCategory() = PlaceCategory(
            id = getLong(getColumnIndex(DbContract.PlaceCategories._ID)),
            name = getString(getColumnIndex(DbContract.PlaceCategories.NAME))
    )
}